Fix Issue 14854 - @disable this inconsistent between structs and classes

tibi77 opened the pull request:
I have created the new deprecation message but I have had let the error message for the structs unchanged.

jacob-carlborg committed:
Add base class for all AST nodes

This is useful to be able to handle all AST nodes the same way, i.e.
in an array. Common code for all AST nodes can also be placed here.

RazvanN7 committed:
Turn deprecation into error for privately imported symbols inside aggregate declarations

wilzbach committed:
Setup Azure Pipelines for DMD

Commit:
Fix Issue  19569 Fixed overload resolution giving priority to covariants when should be ambiguous.

tibi77 committed:
Fix Issue 14854 - @disable this inconsistent between structs and classes
Fixed the identation

tibi77 committed:
Fix Issue 14854 - @disable this inconsistent between structs and classes

wilzbach commented:
Please mind the test failures and add a new test for your modifications.

jacob-carlborg commented on src/dmd/dsymbolsem.d:
```diff
@@ -2894,6 +2894,8 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             printf("sc.parent = %s, parent = %s\n", sc.parent.toChars(), funcdecl.parent ? funcdecl.parent.toChars() : "");
             printf("type: %p, %s\n", funcdecl.type, funcdecl.type.toChars());
         }
+	if (funcdecl.fbody && (funcdecl.storage_class & STC.disable))
```
The indentation is wrong.

tibi77 replied:
This is why the checks are failing?

jacob-carlborg replied:
No, an example of a failing build:

```
std/datetime/systime.d(380): Deprecation: constructor `std.datetime.systime.Clock.this` cannot be annotated with `@disable` because it has a body
```

jacob-carlborg replied:
> This is why the checks are failing?

Yes, that's may be the case: ` Error - file 'dmd/dsymbolsem.d' contains tabs at line 2897`.

There seems to be (at least) two different errors.

tibi77 replied:
But for the deprecation in the "std/datetime/systime.d(380):" this is what the issue suggested to do; "So, I suggest that we deprecate having bodies on _all_ @disabled functions and move to making it an error like it is with @disabling the init property on a struct."  [1]. And I find the idea legit.

[1] https://issues.dlang.org/show_bug.cgi?id=14854

wilzbach replied:
FYI the current `master` version of Phobos needs to compile without deprecation messages, so you would need to fix Phobos first, before introducing deprecation messages in it.

tibi77 replied:
Ok I will work on a fix there. Thank you for the information.

tibi77 replied:
I do not understand why the checks are still giving the error for the deprecation message for "
std/internal/cstring.d(188): Deprecation: generated function `std.internal.cstring.TempCStringBuffer!char.TempCStringBuffer.__aggrPostblit` cannot be annotated with `@disable` because it has a body" even though I have already change the main library and the PR were merged.

wilzbach replied:
Did you repush to your PR after the change was merged?

tibi77 replied:
In phobos?

tibi77 replied:
So. Should u change the dlang tools as well?

tibi77 replied:
All done! Should I commit again?

thewilsonator committed:
Merge pull request #9394 from RazvanN7/Dep_error

Turn deprecation into error for privately imported symbols inside aggregate declarations

wilzbach committed:
Merge pull request #9398 from wilzbach/azure

Setup Azure Pipelines for DMD

Commit:
Tweak the module scapetable declaration to match the cpp header decls

rainers committed:
fix appveyor build: ugly escaping no longer necessary

Commit:
fix issue 19718 - DMD 2.085 changes to gsroa make DCD crashing

Commit:
Merge pull request #9421 from rainers/fix_appveyor

fix appveyor build: ugly escaping no longer necessary
merged-on-behalf-of: Sebastian Wilzbach <sebi.wilzbach@gmail.com>

Commit:
Merge pull request #9422 from Basile-z/issue-19718

fix issue 19718 - DMD 2.085 changes to gsroa make DCD crashing
merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

tibi77 committed:
Fix Issue 14854 - @disable this inconsistent between structs and classes.
Fix identation.

jacob-carlborg committed:
Make most of `dmd.root.rmem` pure

jacob-carlborg committed:
Add attributes to `dmd.root.outbuffer`

WalterBright committed:
partially address Issue 19722 - botched implementation of semantic3Errors causes compiler assert fail

WalterBright committed:
Merge pull request #9425 from WalterBright/unknown-error

partially address Issue 19722 - botched implementation of semantic3Er…

PetarKirov committed:
Merge pull request #9424 from jacob-carlborg/pure-rmem

Make most of `dmd.root.rmem` pure

jacob-carlborg committed:
Add attributes to `dmd.root.aav`

jacob-carlborg committed:
Add attributes to `dmd.root.array`

PetarKirov committed:
Merge pull request #9429 from jacob-carlborg/outbuffer-attributes

Add attributes to `dmd.root.outbuffer`

thewilsonator committed:
Merge pull request #9430 from jacob-carlborg/root-aav-attributes

Add attributes to `dmd.root.aav`

PetarKirov committed:
Merge pull request #9431 from jacob-carlborg/root-array-attributes

Add attributes to `dmd.root.array`

rainers committed:
fix appveyor build: ugly escaping no longer necessary

Commit:
Merge pull request #9432 from rainers/fix_appveyor2

fix appveyor build: ugly escaping no longer necessary
merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

Commit:
Fix Issue 19563 Fix extern(C++) Incorrect ABI passing small struct

Commit:
Merge pull request #9420 from Basile-z/doc-escape-interface

Tweak the module scapetable declaration to match the cpp header decls
merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

wilzbach committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

Commit:
Merge pull request #9443 from wilzbach/merge_stable

Merge remote-tracking branch 'upstream/stable' into merge_stable
merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

kinke committed:
Tweak parallelization when building test tools

The d_do_test and sanitize_json rules can be processed in parallel.
Additionally, the d_do_test executable can be built while its unittests
are built & run in the background (both take around 8 seconds on my box
with LDC).
These tools are prerequisites for the runnable/compilable/fail tests and
are thus built before running the actual tests.
These changes save a total of about 10 seconds on my box with LDC.

The best parallelization for a full test run is achieved via
`DMD_TESTSUITE_MAKE_ARGS="-j<N>" make start_all_tests`; the CI scripts
have been adapted accordingly.

thewilsonator committed:
Merge pull request #9434 from look-at-me/cpp

Fix  Issue 19563 Fix extern(C++) Incorrect ABI passing small struct

thewilsonator committed:
Merge pull request #9446 from kinke/testTools

Tweak parallelization when building test tools

ibuclaw committed:
dmd.dmodule: Don't call FileName.canonicalName if global.inifilename not set

Commit:
Merge pull request #9449 from ibuclaw/nullpath

dmd.dmodule: Don't call FileName.canonicalName if global.inifilename not set.
merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

Commit:
Merge pull request #9408 from jacob-carlborg/ast-node-base-class

Add base class for all AST nodes
merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

Commit:
Merge pull request #9406 from look-at-me/117

[Reg 2.074.1] Fix Issue  19569 Fixed overload resolution giving priority to covariants when should be ambiguous.
merged-on-behalf-of: Razvan Nitu <RazvanN7@users.noreply.github.com>

ibuclaw committed:
fix Issue 19735 - Error: variable extern symbols cannot have initializers

ibuclaw committed:
fix Issue 19734 - isDataseg returns true for non-static declarations

Commit:
Merge pull request #9451 from ibuclaw/issue19734

fix Issue 19734 - isDataseg returns true for non-static declarations
merged-on-behalf-of: Petar Kirov <ZombineDev@users.noreply.github.com>

tibi77 committed:
Hope the tests passes
